app/controllers/api/v1/customers_controller.rb

```diff
@@ -80,9 +80,12 @@ def search_customers
       end
 
       def visible_customers
-        current_api_user.customers.or(
-          Customer.where(enterprise_id: editable_enterprises)
-        )
+        Customer.of(managed_enterprise_ids)
+      end
+
+      def managed_enterprise_ids
+        @managed_enterprise_ids ||= Enterprise.managed_by(current_api_user).
+          select('enterprises.id')
```

Comment:
You are missing the `current_api_user.customers` part. That is the customer records of the user shopping at other enterprises. It's not mentioned in the issue, except that the issue says:

> API endpoints should mimic app behaviour in terms of authentication.

I think that the app allows to change `allow_charges` as user. `terms_and_conditions_accepted_at` is also done by the user during checkout. But I think that it's okay to leave this out and keep your version. Let's just be sure that we make a transparent decision about this because the API shouldn't change in the future.

@lin-d-hop What do you think?

Reply:
Sorry, you're right.
When I read `current_api_user.customers`, I thought that `customers` are defined through the `enterprises` that belong to the `user`. 

Now, I can see that there's a direct link between `customers` and `users`.

Reply:
Just to confirm:
Is it possible for a user U1 to have a customer, and that customer doesn't belong to an enterprise managed by the U1?

From the `Admin::CustomersController`, it looks like this is possible only when the user is an admin.
```ruby
    def user_can_create_customer?
      spree_current_user.admin? ||
        spree_current_user.enterprises.include?(@customer.enterprise)
    end
```

Reply:
If the answer to my question is yes.
I don't this that the code will miss `current_api_user.customers`.

if the user is an `Admin`, `Enterprise.managed_by(current_api_user)` will return all the existing enterprises.
```ruby
  scope :managed_by, lambda { |user|
    if user.has_spree_role?('admin')
      where(nil)
    else
      joins(:enterprise_roles).where('enterprise_roles.user_id = ?', user.id)
    end
  }
```
From the customer model, every customer must belong to an enterprise.
```ruby
  belongs_to :enterprise, optional: false
```
So if the current user is an admin, he'll get all the customers.
Otherwise, he'll get the customer of the enterprises managed by him.

Reply:
@abdellani I think you are right though I'm not exactly sure I follow.

So a user can be a customer and a user can manage an enterprise that has customers. We want to focus on the relationships of the latter. 

> We want to mimic the behaviour of admin/customers in the API.

So an API user should be able to access and edit the customers of the enterprises that they own and manage (ie set via Enterrprise Settings/Users). They should not be able to access and edit customers for enterprises for which their enterprises have permissions to manange (ie set via enterprise permissions).

SuperAdmin can see everyone and crash the system, as always :sweat_smile: 

> I think that the app allows to change allow_charges as user. terms_and_conditions_accepted_at is also done by the user during checkout. But I think that it's okay to leave this out and keep your version. 

Yes we should not be able to edit allow_charges or terms_and_conditions_accepted_at via the API. An enterprise should not easily be able to edit these.

Reply:
I think I added some confusion to this. The association `user.customers` is the customer records of a shopper with the enterprises they bought from. It's not connected to admin at all and I think it's okay to leave it out here. I just raised it because an earlier version allowed to edit those records as well. But I don't think that we want shoppers to be able to update their customer records via the API. The use cases here are for enterprise managers, not customers/shoppers themselves.